ReleaseProcessScript.New/CommandLineGitClient.cs

```diff
@@ -0,0 +1,215 @@
+  {
+    if (string.IsNullOrEmpty(tagName))
+      throw new ArgumentException("Input tag was empty, abort pushing tag.", nameof(tagName));
+
+    if (!TagExists(tagName))
+      throw new InvalidOperationException($"Tag with name {tagName} does not exist, abort pushing tag.");
+
+    var output = ExecuteGitCommandWithOutput($"push origin {tagName}");
+    return output.Success;
+  }
+
+  public bool BranchExists (string branchName)
+  {
+    var output = ExecuteGitCommandWithOutput($"show-ref --verify refs/heads/{branchName}");
+
+    return output.Output.Length > 0;
+  }
+
+  public bool RemoteBranchExists (string remoteName, string branchName)
+  {
+    var output = ExecuteGitCommandWithOutput($"ls-remote --heads {remoteName} {branchName}");
+
+    return output.Output.Length > 0;
+  }
+
+  public bool TagExists (string tagName)
+  {
+    var output = ExecuteGitCommandWithOutput($"show-ref --verify refs/tags/{tagName}");
+
+    return output.Output.Length > 0;
+  }
+
+  public bool IsBranchMerged (string branchName, string possiblyMergedBranchName)
+  {
+    ExecuteGitCommandWithOutput($"checkout {branchName}");
+    var mergedBranches = ExecuteGitCommandWithOutput("branch --merged");
+
+    return mergedBranches.Output.Contains(possiblyMergedBranchName);
+  }
+
+  public void EnsureBranchUpToDate (string branchName)
+  {
+    ExecuteGitCommand($"checkout {branchName}");
+
+    var remoteName = "origin";
+
+    if (!RemoteBranchExists(remoteName, branchName))
+      throw new ArgumentException($"Remote branch of {branchName} does not exist");
+
+    ExecuteGitCommand($"fetch {remoteName} {branchName}");
+
+    var local = ExecuteGitCommandWithOutput($"rev-parse {branchName}").Output;
+    var remote = ExecuteGitCommandWithOutput($"rev-parse {remoteName}/{branchName}").Output;
+    var basis = ExecuteGitCommandWithOutput($"merge-base {branchName} {remoteName}/{branchName}").Output;
+
+    if (local.Equals(remote))
+    {
+      //Up-To-Date. OK
+    }
+    else if (local.Equals(basis))
+    {
+      throw new Exception($"Need to pull, local {branchName} branch is behind on repository {remoteName}");
+    }
+    else if (remote.Equals(basis))
+    {
+      //Need to push, remote branch is behind. Ok
+    }
+    else
+    {
+      throw new Exception($"{branchName} diverged, need to rebase at repository {remoteName}");
+    }
+    
+  }
+
+  public bool PushBranch (string branchName)
+  {
+    ExecuteGitCommand($"checkout {branchName}");
+
+    var remoteName = "origin";
+
+    if (!string.IsNullOrEmpty(remoteName))
+    {
+      var upstream = "";
+      var remoteNameOfBranch = GetRemoteOfBranch(branchName);
+
+      if (string.IsNullOrEmpty(remoteNameOfBranch))
+        if (remoteNameOfBranch.Equals(remoteName))
+          upstream = "-u";
+
+      return ExecuteGitCommandWithOutput($"push {upstream} {remoteName} {branchName}").Success;
+    }
+
+    return false;
+  }
+
+  public string GetRemoteOfBranch (string branchName)
```

Comment:
Reflecting on the last comment, it might be good to change the return type to `string?` here and return null instead of throwing an exception if no remote branch is available. This also makes the code above viable .